### PR TITLE
Image: Improve focus management in lightbox

### LINKED ISSUE #55414: Image: fix focus management and accessibility of the Lightbox
### Description

Follow-up to https://github.com/WordPress/gutenberg/pull/55212

https://github.com/WordPress/gutenberg/pull/55212 improved the Lightbox implementation by changing the button that overlid the entire image to a smaller button placed at the top right of the image. That's a welcome change that solves a couple important issues. It also introduced a different implementation of focus handling of the modal dialog that needs to be fiexed.

Please refer the inline review on https://github.com/WordPress/gutenberg/pull/55212
Quoting from the last comment there:

- The implementation based on `event.pointerType` is buggy, as Chrome and Opera return an empty string (as expected) when using the keyboard while Firefox and Safari return `undefined`. 
- Regardless, differentiating the behavior depending on pointing device vs. 'other' is not OK for accessibility. The modal dialog focus management should always have the same behaevior regardless of mouse, touch, keyboard, other.
- The assumption that users are _always_ either using a pointing device or other is incorrect. For example, I may want to open the modal dialog with a click and close it by pewssing the Escape key.

Depending on the flow used to open the dialog _and_ on the browser in use, this is what I observed when testing:

- When opening the dialog, focus is not transferred to the dialog. As such:
  - Focus stays on the page behind the dialog. Tabbing will navigate the page content behing the modal dialog.
  - The dialgo is not announced by screen readers because focus is elsewhere.
  - The dialog doesn't close on Escape.
  - When closing the dialog, focus should always return to the trigger button, instead this is not the case.
  - As such, there is a full focus loss: tabbing again will restart the tab order from the document root.
- The trigger button doesn't visually expose its accessible name. This is a broadeer issue with all icon buttons ued on the front end and should be addressed holistically.
- The trigger button is only 24 by 24 pixels. While this is compliant with the current WCAG 2.2, I would like to see the button target size a little larger.
- When opening the modal dialog, Initial focus should go to the element with role=dialog instead of the Close button.
- The clickable image is problematic. I'm not sure why we want to make a HTML element have a non-native behavior. I'm not sure that would make it more usable in the first place, as it is a non expected beheavi, especially considering the following point.  It is even more problematic on mobile.
- Still, there is no visual hint that the image can be enlarged. It looks like a normal image and only when hovered or focusd shows a visual hint. As a user, looking at a page with a Lightbox image, I wouldn't have any clue that the image can be enlarged. See screnshot below: how users are supposed to understand the image can be enlarged?

Screenshot 2023-10-17 at 13 12 47

### Step-by-step reproduction instructions

Testing on macOS:
- Add an image block and set the image to open in the Lightbox.
- Publish and view the front end.
- Makes sure full keyboard navigation is enabled in your macOS settings under Keyboard. See screenshot:

Screenshot 2023-10-17 at 15 42 46

- With macOS Chrome:
  - Tab to the button that opens the image modal dialog.
  - Press Enter or Spacebar.
  - Observe that focus is moved to the modal dialog Close button, as expected.
- With macOS Firefox:
  - Repeat the steps above.
  - Observe focus is _not_ moved to the modal dialog.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by artemiomorales**
## What?

This PR makes the focus management more consistent in the image lightbox by removing the reference to `event.pointerType` and the differentiation in behavior when using a mouse versus a keyboard.

## Why?

Addresses https://github.com/WordPress/gutenberg/issues/55414 and improves on changes made in https://github.com/WordPress/gutenberg/pull/55212.

Altogether, the previous approach was unreliable, so this PR aims to fix that.

## How?

It removes the logic using `event.pointerType`, and instead...
1.  Makes sure focus is always set to the close button when the lightbox opens.
2. Adds a `tabindex` of -1 to the image, which prevents the focus ring from erroneously showing around the close button when the lightbox is opened in Chrome and Firefox (the focus ring appears to always be visible in this case in desktop Safari, which appears to be the intended browser behavior).
3. Stores a reference to the trigger button so that we can set focus to it when the lightbox closes.

## Testing Instructions

1. Add an image to a post
3. Enable 'Expand on Click' in the image's block settings
4. Publish and view the post
5. Hover over the image and see that the cursor becomes a magnifying glass and that a button appears in the top right corner
6. Click on the image or the button to open the lightbox

On keyboard, test that tabbing to the image will show the lightbox trigger button on focus and that pressing `Enter` or `Spacebar` opens the lightbox as expected.

While the lightbox is open, test that...
1. Pressing `Escape` closes the lightbox
2. Pressing `Tab` will move focus to the close button and trap it
3. That pressing `Enter` or `Spacebar` while on the close button will close the lightbox
4. Clicking on the modal or the close button will close the lightbox

When the lightbox closes, make sure focus is set on the lightbox trigger.

### CODE REVIEW

**Comment by artemiomorales** [comment]
One issue with setting the focus on the dialog itself as recommended in https://github.com/WordPress/gutenberg/issues/55414 is that this causes unexpected behavior in Safari. Usually it will cause VoiceOver to _say_ that focus is on the close button, but it will not respond to an `Enter` keypress as expected. This may be because the close button is a child of the dialog, and having focus on the parent isn't supported.

I've recorded a video showing this behavior while casting the keystrokes to the screen:

https://github.com/WordPress/gutenberg/assets/5360536/572f8cf7-0aa8-481c-9cb7-439281165bfa

**Comment by artemiomorales** [comment]
I'm ready to get additional eyes on this and suggestions on how to improve the implementation. As of this writing, I've been unable to solve [this issue](https://github.com/WordPress/gutenberg/pull/55428#issuecomment-1767230218) in Safari while setting focus on the modal itself, i.e the element with`role=dialog`.

**Comment by artemiomorales** [comment]
@afercia After looking at this with @SantosGuillamot, I don't think there's a way to set focus to the whole modal dialog without causing unexpected behavior ([see video above](https://github.com/WordPress/gutenberg/pull/55428#issuecomment-1767230218)).

So with that in mind, I've set the focus back to the close button, and I've also added a `tabindex` to the image, which at least prevents the focus ring from unnecessarily showing up around the close button on Chrome and Firefox when a mouse is used to open the lightbox (although it seems the focus ring is always visible in desktop Safari). The keyboard navigation seems to work in all cases and still displays the focus ring as expected.

I know you're not a fan of making the image itself clickable, but are there any technical drawbacks to adding a `tabindex` of -1 to the image or other issues this could cause?

File: packages/block-library/src/image/index.php [hunk 1]
```diff
@@ -218,6 +218,7 @@ function block_core_image_render_lightbox( $block_content, $block ) {
 		)
 	);
 	$w->next_tag( 'img' );
+	$w->set_attribute( 'tabindex', '-1' );
```
Comment (afercia) [inline packages/block-library/src/image/index.php]: After the last two commits, I think thie tabindex can be removed.

**Comment by afercia** [comment]
> @afercia After looking at this with @SantosGuillamot, I don't think there's a way to set focus to the whole modal dialog without causing unexpected behavior ([see video above](https://github.com/WordPress/gutenberg/pull/55428#issuecomment-1767230218)).

> One issue with setting the focus on the dialog itself as recommended in https://github.com/WordPress/gutenberg/issues/55414 is that this causes unexpected behavior in Safari. Usually it will cause VoiceOver to say that focus is on the close button, but it will not respond to an Enter keypress as expected. This may be because the close button is a child of the dialog, and having focus on the parent isn't supported.

@artemiomorales thanks for looking into this.
There's an incorrect assumption here that is probably a source of confusion. With Safari and VoiceOver, the correct way to press a button is by using Control+Option+Spacebar. Enter will only work if focus is on the button. But, actually, focus in on the dialog.
What happens here is specific to Safari and VoiveOver I think it is ok. When the modal dialog opens:
- Keyboard focus is actually on the dialog.
- The VoiceOver 'virtual cursor' is on the Close button instead. 
- Control+Option+Spacebar will close the dialog, as expected.
- Enter will not close the dialog, because keyboard focus in on the dialog element.
- If you press Tab, you will hear VoiceOver announces the Close button (again). This proves that focus was on the dialog and after pressing Tab goes to the button.
- At this point, pressing Enter will work.

This is expected behavior.

There is one thing that behaves differently compared to the editor Mocal component though.
When the modal dialog opens, VOiceOver should not move virtual cursor to the Close button in the first place. This doesn't happen for example on the editor preferences modal dialog.
I have no idea why this different beheavior occurs but there are markup differences and on the Lightbox we are setting some aria-* attributes dynamically. That could have an impact. In any case, I'd recomment to focus the dialog. See https://github.com/WordPress/gutenberg/pull/55428/commits/ab653a4acea9e1b6b8a20d0fe303f1f2ecc22d1e that reinstates your first implementation and adds a negative tabindex to teh fialog to make it focusable. This is consistent with the Editor Modal component.

As an aside and out of curiosity, I spent some time trying to debug the Safari + VoiceOver behavior. I think it's something specific to Safari and the aria-modal attribute.
- Comment out the `ref.focus();` that sets focus on the dialog. 
- Test with only Safari, keep VoiceOver off.
- Open the dialog with the keyboard Enter or Spacebar keys.
- Observe that Safari sets focus on the Close button. This is weird because there is nothing that moves focus on the Close button. It appears to be a Safari thing. To me, it's a bug but I'm pretty sure the Apple fols would call it a "feature".
- To confirm that focus in on the Close button, press Enter or Spacebar  and observe the dialog closes.
- Test with other browsers and observe that focus is not moved to the dialog, as expected, because we removed any focus management.
- In the php file, remove the line to handle the aria-modal attribute `data-wp-bind--aria-modal="selectors.core.image.ariaModal"`.
- Test again with only Safari.
- Open the dialog with the keyboard Enter or Spacebar keys.
- Observe that this time focus is not moved to the dialog, as expected.

So it appears that Safari, when it sees an aria-modal attribute, now moves focus to the first focusable element within the element with aria-modal. To me, this is a bug. It may be related to recent work that browser vendors have been doing to align their implementations with the new focus steps for the HTML `<dialog>` element. Reference:
https://github.com/whatwg/html/wiki/dialog--initial-focus,-a-proposal
https://github.com/whatwg/html/pull/4184

**Comment by afercia** [comment]
@aristath @carolinan and myself had a look at this and we think it's the smallest amount of changes we can introduce since we are in RC1. We recommend final review and testing with all major browsers.

@joedolson @alexstine I'd appreciate your thoughts and testing, especially with Windows screen readers. Thanks.

There are outstanding, broader, issues that would require some discussion and this is not the right time for that, as we are in RC1:
- Absence of any visual hint that an image can be enlarged. In a page with many images, where only a few of them open in a Lightbox, users wan't have any way to understand the difference unless they hover or focus an image. we are not sure this is a good design and UX in the first place. 
- The image element is clickable and it opens the dialog. This is a non expected interaction and should be reconsidered. Windows screen readers will probably announce the image as 'clickable'.
- The trigger button should enforce a minimum size of 24 by 24 pixels.
- The Lightbox does not have feature parity with the Navigation, where users can change the appearance of the buttons.
- The icon buttons don't visually expose their accessible name.

We agree all these points can't be addressed now. The should be addressed soon in the next release though.

**Review by afercia — COMMENTED** [review]
Pending final review by @artemiomorales this LGTM

**Review by alexstine — APPROVED** [review]
I don't have time to test this at the moment but the code changes look good until it can be revisited during next release.

File: packages/block-library/src/image/index.php (hunk 1 above)
Comment (alexstine) [reply packages/block-library/src/image/index.php]: Agreed, that looks to be no longer needed.

**Review by jameskoster — APPROVED** [review]
It's working as advertised for me and is an improvement compared to trunk.

**Review by michalczaplinski — APPROVED** [review]
Tested it according to the instructions and it works as expected for me. 

I don't have the a11y expertise to bring much nuance to the discussion but happy with the change since it's an incremental improvement over the current state of the Lightbox.

File: packages/block-library/src/image/index.php (hunk 1 above)
Comment (artemiomorales) [reply packages/block-library/src/image/index.php]: Yes the PR seems to work well without it, so I've removed it.

**Comment by artemiomorales** [comment]
> So it appears that Safari, when it sees an aria-modal attribute, now moves focus to the first focusable element within the element with aria-modal. To me, this is a bug. It may be related to recent work that browser vendors have been doing to align their implementations with the new focus steps for the HTML <dialog> element. Reference:
https://github.com/whatwg/html/wiki/dialog--initial-focus,-a-proposal
https://github.com/whatwg/html/pull/4184

@afercia Thank you for your detailed analysis and exploration. The approach we have now looks good to me. I've pushed a change removing the extra `tabindex` and will merge once the tests pass.

**Comment by afercia** [comment]
Thanks everyone for the PR, review, and feedback. From an a11y perspective, this looks good to me though I'd like to see it tested with WIndows screen readers.

We can investigate the different behavior of Safari / VoiceOver with the Lightbox modal dialog compared to the editor Modal component later in teh next release cycle.

**Comment by afercia** [comment]
I quickly tested on [gutenberg.run](http://gutenberg.run/) with latest Firefox and NVDA. 
As I was thinking, the image is announced as `graphic clickable`. Other than that, interaction and announcement with a screen reader look good to me. I think this looks good for RC 2. Screenshot: 

Screenshot 2023-10-20 165058

Notice the doubled images within the dialog are both announced.

I will create a follow-up issue to evaluate improvements for the pending points.

**Review by alexstine — APPROVED** [review]
@artemiomorales Tested on Windows with NVDA in Google Chrome and Firefox. This works well enough for now. Let's land it.

**Comment by artemiomorales** [comment]
Great, thank you all for the feedback and testing! This has been merged and will follow up on outstanding items in the next release cycle.

**Comment by SiobhyB** [comment]
Cherry-picked for inclusion in 6.4 in https://github.com/WordPress/gutenberg/pull/55481/commits/feefbe15bab147e93b196aaef5c3cef40e23445a.

**Pull request merged**